### PR TITLE
PXD-2213 Feat/empty record

### PR DESCRIPTION
**Pull request by giangbui**
New endpoint for blank records containing just the `uploader` field

- Allow creating empty records
- Query by empty acl
- Support filling (still not modifying) hash

### CODE REVIEW

File: indexd/index/drivers/alchemy.py [hunk 1]
```diff
@@ -633,6 +635,71 @@ def add(self,
 
             return record.did, record.rev, record.baseid
 
+    def add_blank_record(self, did=None, uploader=None, baseid=None):
```
Comment (rudyardrichter) [inline indexd/index/drivers/alchemy.py]: `uploader` should not be optional for this flow, and `baseid` should not be provided

Comment (rudyardrichter) [inline indexd/index/drivers/alchemy.py]: also `did` should not be provided for this case, indexd should generate a new GUID and send it to fence

File: indexd/index/drivers/alchemy.py [hunk 2]
```diff
@@ -633,6 +635,71 @@ def add(self,
...
+            
+            session.merge(base_version)
+            session.add(record)
+            session.commit()
+
+            return record.did, record.rev, record.baseid
+    
+    def update_blank_record(self, did, rev, size=None, hashes=None):
+        """
+        Create a new blank record with only
```
Comment (rudyardrichter) [inline indexd/index/drivers/alchemy.py]: docstring needs to be updated

File: indexd/index/drivers/alchemy.py [hunk 3]
```diff
@@ -633,6 +635,71 @@ def add(self,
...
+        """
+        if hashes is None:
```
Comment (rudyardrichter) [inline indexd/index/drivers/alchemy.py]: can onelinerize: `hashes = hashes or {}` :slightly_smiling_face:

File: indexd/index/drivers/alchemy.py (hunk 1 above)
Comment (giangbui) [reply indexd/index/drivers/alchemy.py]: Fixed in 83f5b6b6dfc04b9eaa7d6502d320ae6c639ba1aa

File: indexd/index/drivers/alchemy.py (hunk 2 above)
Comment (giangbui) [reply indexd/index/drivers/alchemy.py]: Fixed in 83f5b6b6dfc04b9eaa7d6502d320ae6c639ba1aa

File: indexd/index/drivers/alchemy.py (hunk 3 above)
Comment (giangbui) [reply indexd/index/drivers/alchemy.py]: Fixed several other places in 83f5b6b6dfc04b9eaa7d6502d320ae6c639ba1aa

File: openapis/swagger.yaml [hunk 4]
```diff
@@ -236,6 +241,68 @@ paths:
...
+          description: Metadata object that needs to be added to the store
+          required: true
+          schema:
+            $ref: '#/definitions/InputBlankInfo'
+      responses:
+        '200':
+          description: successful operation
+          schema:
+            $ref: '#/definitions/OutputInfo'
+        '400':
```
[the remark this answers is not in the document]
Comment (giangbui) [reply openapis/swagger.yaml]: Changed some code blocks to raise exceptions when necessary

File: README.md [hunk 5]
```diff
@@ -31,7 +31,14 @@ Using AWS SNS or Google PubSub it is possible to have streaming notifications wh
...
 For existing data in buckets, the SNS or PubSub notifications may be simulated such that the indexing functions are started for each object in the bucket. This is useful because only a single code path is necessary for indexing the contents of an object.
 
-3. Using the Indexd REST API for record insertion.
+3. Indexing void object for fully control the bucket structure.
+
+Indexd supports void or blank records that allows users to pre-register data files in indexd before actually registering them. The complete flow contains three main steps: pre-register, hash/size/url populating and data node registration:
+    - Fence requests blank object from indexd. Indexd creates an object with no hash, size, and urls except the `uploader` field.
+    - Indexd listener mornitors bucket update, update to indexd with url, hash, size.
+    - Windmill lists indexd records that’s owned by the user that has empty `acl`. The user fills all empty fields and submit the request to indexd to update the `acl`.
+
```
Comment (rudyardrichter) [inline README.md]: could you add a link to the data upload docs? https://github.com/uc-cdis/cdis-wiki/tree/master/dev/gen3/data_upload

File: README.md [hunk 6]
```diff
@@ -31,7 +31,14 @@ Using AWS SNS or Google PubSub it is possible to have streaming notifications wh
 
 For existing data in buckets, the SNS or PubSub notifications may be simulated such that the indexing functions are started for each object in the bucket. This is useful because only a single code path is necessary for indexing the contents of an object.
 
-3. Using the Indexd REST API for record insertion.
+3. Indexing void object for fully control the bucket structure.
+
+Indexd supports void or blank records that allows users to pre-register data files in indexd before actually registering them. The complete flow contains three main steps: pre-register, hash/size/url populating and data node registration:
+    - Fence requests blank object from indexd. Indexd creates an object with no hash, size, and urls except the `uploader` field.
+    - Indexd listener mornitors bucket update, update to indexd with url, hash, size.
+    - Windmill lists indexd records that’s owned by the user that has empty `acl`. The user fills all empty fields and submit the request to indexd to update the `acl`.
```
Comment (rudyardrichter) [inline README.md]: this point on `windmill` is subject to change as we develop this—I would say something along the lines of `the client application (windmill or gen3-data-client) lists records for data files which the user needs to submit to the graph`.

File: indexd/index/blueprint.py [hunk 7]
```diff
@@ -264,6 +266,57 @@ def post_index_record():
...
 
 
+@blueprint.route('/index/blank/', methods=['POST'])
+@authorize
+def post_index_blank_record():
+    '''
+    Create a blank new record with only uploader field is filled
+    '''
+
+    uploader = flask.request.json.get('uploader')
```
Comment (rudyardrichter) [inline indexd/index/blueprint.py]: `request.get_json()` encouraged over `.json`

File: indexd/index/blueprint.py [hunk 8]
```diff
@@ -264,6 +266,57 @@ def post_index_record():
...
+        uploader=uploader
+    )
+
+    ret = {
+        'did': did,
+        'rev': rev,
+        'baseid': baseid,
+    }
+
+    return flask.jsonify(ret), 200
```
Comment (rudyardrichter) [inline indexd/index/blueprint.py]: I think a 201 status code (`CREATED`) would make the most sense, though the other indexd endpoints already are just 200, so if you want to leave this alone that's fine

File: README.md (hunk 6 above)
Comment (giangbui) [reply README.md]: Thanks. Fixed in 7caea53ee872202df117db75459f21526c7b2db1

File: indexd/index/blueprint.py (hunk 7 above)
Comment (giangbui) [reply indexd/index/blueprint.py]: Fixed in 7caea53ee872202df117db75459f21526c7b2db1

File: indexd/index/blueprint.py (hunk 8 above)
Comment (giangbui) [reply indexd/index/blueprint.py]: Good point!. Fixed in 7caea53ee872202df117db75459f21526c7b2db1

**Pull request merged**